core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -1272,6 +1274,138 @@ private static boolean isWidening(RelDataType type, RelDataType type1) {
     return register(rel, newJoin, mapOldToNewOutputs, corDefOutputs);
   }
 
+  public @Nullable Frame decorrelateRel(LogicalUnion rel, boolean isCorVarDefined) {
+    return decorrelateRel((Union) rel, isCorVarDefined);
+  }
+
+  public @Nullable Frame decorrelateRel(Union rel, boolean isCorVarDefined) {
+    assert !cm.mapRefRelToCorRef.containsKey(rel);
+
+    final List<RelNode> oldInputs = rel.getInputs();
+
+    final List<Frame> inputFrames = new ArrayList<>();
+    final Set<CorDef> corDefSet = new HashSet<>();
+
+    // Decorrelate to collect input frames and collect all
+    // the corDefOutputs from the frames. Note that the
+    // corDefOutputs might be different among the frames.
+    for (RelNode oldInput: oldInputs) {
+      final Frame frame = getInvoke(oldInput, isCorVarDefined, rel);
+      if (frame == null) {
```

Comment:
What if one part  of the union is changed due to de-correlation and other part is not changed ? why do we return null here even if one part of the union has changed? 
Eg - one side has Value Rel (hence returns null) , other has Correlate-...-TableScan